### DISCUSSION TITLE
Nested lambda syntax sugar for procedure definitions

### DISCUSSION THREAD

**Discussion by bcalco** [leppie/IronScheme#88] [Discussions]
@leppie:

MIT Scheme and Racket both support "nested" lambda syntax sugar for defining procedures. (Curiously, Chez does not, even though the latest Racket is built on Chez.)

For example, instead of being "stuck" at one-level syntax sugar for the following definition:

```scheme
(define (compose f g)
  (lambda (x)
    (f (g x))))
```
It's possible to "nest" or "roll up" the lambda definition with "extra" syntax sugar:

```scheme
(define ((compose f g) x)
  (f (g x)))
```

While I am not particularly hung up on this approach personally (I actually prefer to type out the nested lambdas explicitly, even for one-level lambda procedures), my users have asked about using code from several Scheme books that do use this syntax, and some definitions won't compile. My application is for the mathematically inclined, so I guess I should have foreseen that they might want to use code from popular Scheme books.

Anyway, I was wondering if there is any chance you plan or would be willing to add support in IronScheme for this extra syntax sugar for procedure definitions? I did not add this to "issues" as I figure you'd rather talk it out and decide whether it's an issue yourself.

If it's not too heavy a lift, I would argue IronScheme should support the syntax, only because it's used in these well-known books and supported by at least two other popular Scheme implementations. I recognize some Scheme implementations get on happily without it (IronScheme including), but my application, which embeds IronScheme, should welcome code from such books, if possible. I have too much else to work on in my application at the moment to want to go down that rabbit hole right now, and since it seems like a general Scheme feature of some utility, I thought I'd find out first if you could do it.

If you're wondering which Scheme books I'm referring to, they are:

[1] Functional Differential Geometry (2013, MIT Press)

[2] Structure and Interpretation of Classical Mechanics, 2nd Ed. (2014, MIT Press)

[3] Software Design for Flexibility (2021, MIT Press)

**Comment by leppie** [comment]
I dont know the exact semantics, but what you are showing, can be achieved quite easily with a macro.

```
(define-syntax define*
  (syntax-rules ()
    [(_ ((name a ...) b ...) body ...)
      (define (name a ...) 
        (lambda (b ...)
          body ...))]))
```

```
(define* ((compose f g) x)
  (f (g x)))

((compose abs +) -1) ;=> 1
```

Obviously, if you want to replace `define` in your library/environment, you would need to 'hide' the original `define` and implement cases for all the normal `define` forms.  Hence why I just called it something else to illustrate, and just one level, but to define it recursively should not be too hard.

**Comment by lassik** [comment]
Standardized in [SRFI 219: Define higher-order lambda](https://srfi.schemers.org/srfi-219/)

**Reply by leppie** [reply]
I was going to suggest "someone must have done it" :D 

I think just adding the SRFI will make most sense.

**Reply by bcalco** [reply]
Indeed, [here it is](https://github.com/scheme-requests-for-implementation/srfi-219/blob/master/srfi/219.sls).

```scheme
(library (srfi :219)
  (export define)
  (import (rename (rnrs base) (define native-define)))
  (define-syntax define
    (syntax-rules ()
      ((define ((head . outer-args) . args) . body)
       (define (head . outer-args) (lambda args . body)))
      ((define head . body)
       (native-define head . body)))))
```

**Reply by bcalco** [reply]
I wasn't aware the SRFI was standardized so recently, but I'm not surprised it was so easy to implement. Macros freaking rock. Will give it a spin locally but I don't suspect it will be any problem to use.  Is it something you plan to add to the srfi submodule in light of this discussion, @leppie?

**Reply by lassik** [reply]
> Macros freaking rock.

Not only macros, but Scheme's library system is an unsung hero! In how many languages can you rename the native `define`, define your own `define` in terms of it, and export the new `define` for seamless use in other modules? The combined library and macro expander is incredible.

**Reply by lassik** [reply]
The fact that libraries and hygienic macros are rolled up together into the same expander, and _that's how it should be_, is one of the major revelations awaiting intermediate level schemers, after they have got over the shock of how seamlessly pervasive tail calls and continuations work together.

**Reply by leppie** [reply]
> Is it something you plan to add to the srfi submodule in light of this discussion, @leppie?

IronScheme tracks the chez srfi libraries, so I would need to check. Think it was up to around srfi 100, so many to add and test and see what can work or not or need fixes.

**Comment by leppie** [comment]
I have setup an initial PR for merging upstream SRFI. All seem fine except for SRFI 158.

https://github.com/IronScheme/chez-srfi/pull/1

**Comment by leppie** [comment]
I have merged upstream SRFI. New release https://github.com/IronScheme/IronScheme/releases/tag/1.0.243

Still need to add the tests for the new SRFI's.
